### PR TITLE
Update Vulkan RoPE implementation

### PR DESCRIPTION
**Pull request by 0cc4m**
Implement #7634  for Vulkan

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -481,7 +481,7 @@ if (LLAMA_CUDA)
         if (LLAMA_CUDA_F16 OR LLAMA_CUDA_DMMV_F16)
             set(CMAKE_CUDA_ARCHITECTURES "60;61;70") # needed for f16 CUDA intrinsics
         else()
-            set(CMAKE_CUDA_ARCHITECTURES "52;61;70") # lowest CUDA 12 standard + lowest for integer intrinsics
+            set(CMAKE_CUDA_ARCHITECTURES "86") # lowest CUDA 12 standard + lowest for integer intrinsics
```
Comment (slaren) [inline CMakeLists.txt:484]: This shouldn't be changed. I imagine it was set for faster builds while testing, you can achieve the same by configuring with `-DCMAKE_CUDA_ARCHITECTURES=86`.

Comment (0cc4m) [reply CMakeLists.txt:484]: My apologies, that wasn't supposed to be there. It's part of a different problem where I can't run the CUDA code anymore if I don't do that. I plan to open an issue about that.

**Review by slaren — APPROVED** [review]
`test-backend-ops` passes

**Comment by 0cc4m** [comment]
@slaren I added the code to catch the exception in `ggml_backend_vk_buffer_type_alloc_buffer` when the allocation fails and instead return a nullptr.

This leads to a segfault in `alloc_tensor_range` https://github.com/ggerganov/llama.cpp/blob/9fc0f55cd97e12df26543c4e0d0ffecac9934956/ggml-alloc.c#L888-L890
```
ggml_vulkan: Validation layers enabled
ggml_vulkan: Found 1 Vulkan devices:
Vulkan0: NVIDIA GeForce RTX 3090 | uma: 0 | fp16: 1 | warp size: 32
llm_load_tensors: ggml ctx size =    0,83 MiB
ggml_vulkan: Device memory allocation of size 1613332480 failed.
ggml_vulkan: vk::Device::allocateMemory: ErrorOutOfDeviceMemory
alloc_tensor_range: failed to allocate Vulkan0 buffer of size 1613332480
[1]    1021791 segmentation fault (core dumped)  build_vk/bin/main -f ~/llama.cpp/input.txt -c 2048 -n 128 --ignore-eos -m
```
Taking a closer look with a debugger, `n_buffers=6`, `i=1` and `buffers[1]=0x6`. That is protected memory, causing the segfault when trying to derefence it. Any idea how that value got there?

**Comment by slaren** [comment]
This should fix it - it's a dumb mistake in the usage of operator precedence.
```diff
iff --git a/ggml-alloc.c b/ggml-alloc.c
index 73a3c157..eb75962d 100644
--- a/ggml-alloc.c
+++ b/ggml-alloc.c
@@ -886,7 +886,7 @@ static bool alloc_tensor_range(struct ggml_context * ctx,
         fprintf(stderr, "%s: failed to allocate %s buffer of size %zu\n", __func__, ggml_backend_buft_name(buft), size);
 #endif
         for (size_t i = 0; i < *n_buffers; i++) {
-            ggml_backend_buffer_free(*buffers[i]);
+            ggml_backend_buffer_free((*buffers)[i]);
         }
         free(*buffers);
         return false;
```

**Comment by 0cc4m** [comment]
Yeah, that was it. Thanks.

**Pull request merged**